File: CodeForces/Contest/Round #845 (Div 2)/E845.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class E845{
   
   public static int n;
   public static ArrayList<ArrayList<Edge>> adj;
   
   public static HashSet<Integer> seen;
   
   public static Stack<Integer> stack;
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      int t = Integer.parseInt(f.readLine());
      
      for(int q = 1; q <= t; q++){
      
         StringTokenizer st = new StringTokenizer(f.readLine());
      
         n = Integer.parseInt(st.nextToken());
         int m = Integer.parseInt(st.nextToken());
         
         ArrayList<Integer> edges = new ArrayList<Integer>();
         
         adj = new ArrayList<ArrayList<Edge>>(n+1);
         for(int k = 0; k <= n; k++){
            adj.add(new ArrayList<Edge>());
         }
         
         for(int k = 0; k < m; k++){
            st = new StringTokenizer(f.readLine());
                
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            
            adj.get(a).add(new Edge(b,w,1));
            adj.get(b).add(new Edge(a,w,-1));
            
            edges.add(w);
         }
         
         //check if possible (if connected or not)
         seen = new HashSet<Integer>();
         seen.add(1);
         dfsconnected(1,Integer.MAX_VALUE);
         
         if(seen.size() != n){
            out.println(-1);
            continue;
         }
         
         Collections.sort(edges);
         
         int l = 0;
         int r = m-1;
         int ans = edges.get(m-1);
         while(l <= r){
            int mid = l + (r-l)/2;
            
            //thresh is edges.get(mid)
            if(check(edges.get(mid))){
               if(mid == 0) ans = 0;
               else ans = edges.get(mid-1);
               
               r = mid-1;
            } else {
               l = mid+1;
            }
         }
         
         out.println(ans);
         
      }
      
      
      
      
      out.close();
   }
   
   
   
   public static void topsort(int v, int thresh){
      
      for(Edge e : adj.get(v)){
         if(seen.contains(e.to)) continue;
         if(e.dir == -1 && e.w >= thresh) continue;
         seen.add(e.to);
         topsort(e.to,thresh);
      }
      
      stack.push(v);
      
   }
   
   
   public static boolean check(int thresh){
      //do topsort
      stack = new Stack<Integer>();
      seen = new HashSet<Integer>();
      
      for(int k = 1; k <= n; k++){
         if(!seen.contains(k)){
            seen.add(k);
            topsort(k,thresh);
         }
        
      }
      
      //get top of topsort
      int top = stack.pop();
      
      //check if connected
      seen = new HashSet<Integer>();
      seen.add(top);
      dfsconnected(top,thresh);
      return seen.size() == n;
   }
   
   
   public static void dfsconnected(int v,int thresh){
      
      for(Edge e : adj.get(v)){
         if(seen.contains(e.to)) continue;
         if(e.dir == -1 && e.w >= thresh) continue;
         seen.add(e.to);
         dfsconnected(e.to,thresh);
      }
      
      
   }
   
   
   public static class Edge{
      int to;
      int w;
      int dir;
      public Edge(int a, int b, int c){
         to = a;
         w = b;
         dir = c;
      }
        
   }
      
}

File: CodeForces/Contest/Round #845 (Div 2)/E845.cpp
#include <bits/stdc++.h>

using namespace std;

struct Edge{
   int to;
   int w;
   int dir;
};

int n;
vector<vector<Edge>> adj;

unordered_set<int> seen;

stack<int> topstack;


void dfsconnected(int v, int thresh){
   for (const auto& e : adj[v]){
      if(seen.find(e.to) != seen.end()) 
         continue;
      if(e.dir == -1 && e.w >= thresh) 
         continue;
      seen.insert(e.to);
      dfsconnected(e.to,thresh);
   }
}

void topsort(int v, int thresh){
   for( const auto& e : adj[v]){
      if(seen.find(e.to) != seen.end()) 
         continue;
      if(e.dir == -1 && e.w >= thresh) 
         continue;
      seen.insert(e.to);
      topsort(e.to,thresh);
   }
   
   topstack.push(v);
}

bool check(int thresh){
   //do topsort
   topstack = stack<int>();
   seen = unordered_set<int>();
   
   for(int k = 1; k <= n; k++){
      if(seen.find(k) == seen.end()){
         seen.insert(k);
         topsort(k,thresh);
      }
   }
   
   //get top of topsort
   int ttop = topstack.top();
   
   //check if connected
   seen = unordered_set<int>();
   seen.insert(ttop);
   dfsconnected(ttop,thresh);
   return seen.size() == n;
}




int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
      
   int t;
   cin >> t;
   for(int q = 0; q < t; q++){
   
      int m;
      cin >> n >> m;
      
      vector<int> edges;
      
      adj = vector<vector<Edge>>(n+1,vector<Edge>());
      
      for(int k = 0; k < m; k++){
         int a,b,w;
         cin >> a >> b >> w;
         
         Edge e1 = {b,w,1};
         Edge e2 = {a,w,-1};
         adj[a].push_back(e1);
         adj[b].push_back(e2);
         
         edges.push_back(w);
      }
      
      //check if connected
      seen = unordered_set<int>();
      seen.insert(1);
      dfsconnected(1,INT_MAX);
      if(seen.size() != n){
         cout << -1 << endl;
         continue;
      }
      
      sort(edges.begin(),edges.end());
      
      int l = 0;
      int r = m-1;
      int ans = edges[m-1];
      while(l <= r){
         int mid = l + (r-l)/2;
            
         //thresh is edges.get(mid)
         if(check(edges[mid])){
            if(mid == 0) ans = 0;
            else ans = edges[mid-1];
               
            r = mid-1;
         } 
         else {
            l = mid+1;
         }
      }
      
      cout << ans << endl;
   
   }
   
   
   return 0;
}